### ISSUE TITLE
The loss fails to converge when optimizing rotation matrices

### ISSUE THREAD

**Issue by limingwhat** [facebookresearch/SpinQuant#21]
It is really a nice work! But when I tried to run optimize_rotation.py on a small 1.5B model, the loss keeps bouncing up and down and fails to converge. I have tried smaller and larger learning rate, all fail. Is there any guideline I can follow to adjust the training optimization parameters? @liuzechun Very much appreciate your help!

**Comment by limingwhat** [comment]
```
{'loss': 3.0266, 'grad_norm': 44.073368072509766, 'learning_rate': 1.4996299202742986, 'epoch': 0.0}                                                                                                                                                                                      
{'loss': 14.4104, 'grad_norm': 339.9272766113281, 'learning_rate': 1.4985200463212036, 'epoch': 0.0}                                                                                                                                                                                      
{'loss': 15.2535, 'grad_norm': 200.8673858642578, 'learning_rate': 1.4966714734523099, 'epoch': 0.0}
{'loss': 12.7564, 'grad_norm': 201.1876983642578, 'learning_rate': 1.4940860259858586, 'epoch': 0.0}                                                                                                                                          
{'loss': 13.7565, 'grad_norm': 165.16897583007812, 'learning_rate': 1.4907662554463532, 'epoch': 0.0}                                                                                                                                         
{'loss': 12.8208, 'grad_norm': 299.32452392578125, 'learning_rate': 1.4867154380465166, 'epoch': 0.0}                                                                                                                                         
{'loss': 9.8955, 'grad_norm': 409.35687255859375, 'learning_rate': 1.4819375714540606, 'epoch': 0.01}                                                                                                                                         
{'loss': 12.7699, 'grad_norm': 201.9282684326172, 'learning_rate': 1.4764373708464733, 'epoch': 0.01}                                                                                                                                         
{'loss': 11.029, 'grad_norm': 116.45204162597656, 'learning_rate': 1.4702202642577074, 'epoch': 0.01}                                                                                                                                         
{'loss': 11.3271, 'grad_norm': 157.50218200683594, 'learning_rate': 1.4632923872213652, 'epoch': 0.01}                                                                                                                                        
{'loss': 13.0735, 'grad_norm': 186.06764221191406, 'learning_rate': 1.4556605767156692, 'epoch': 0.01}                                                                                                                                        
{'loss': 12.7235, 'grad_norm': 305.3439636230469, 'learning_rate': 1.4473323644161886, 'epoch': 0.01}                                                                                                                                         
{'loss': 10.8752, 'grad_norm': 150.50645446777344, 'learning_rate': 1.4383159692629857, 'epoch': 0.01}                                                                                                                                        
{'loss': 11.3303, 'grad_norm': 212.85255432128906, 'learning_rate': 1.4286202893495146, 'epoch': 0.01}                                                                                                                                        
{'loss': 11.642, 'grad_norm': 198.8216094970703, 'learning_rate': 1.4182548931412757, 'epoch': 0.01}                                                                                                                                          
{'loss': 11.3271, 'grad_norm': 118.17646789550781, 'learning_rate': 1.4072300100328976, 'epoch': 0.01}                                                                                                                                        
{'loss': 11.301, 'grad_norm': 155.64779663085938, 'learning_rate': 1.3955565202529576, 'epoch': 0.01}                                                                                                                                         
{'loss': 9.5373, 'grad_norm': 178.14039611816406, 'learning_rate': 1.3832459441265113, 'epoch': 0.01}                                                                                                                                         
{'loss': 9.6913, 'grad_norm': 170.9600372314453, 'learning_rate': 1.3703104307059213, 'epoch': 0.02}                                                                                                                                          
{'loss': 9.3981, 'grad_norm': 251.22030639648438, 'learning_rate': 1.3567627457812106, 'epoch': 0.02}                                                                                                                                         
{'loss': 11.9772, 'grad_norm': 250.7728271484375, 'learning_rate': 1.3426162592817679, 'epoch': 0.02}                                                                                                                                         
{'loss': 12.2085, 'grad_norm': 278.52642822265625, 'learning_rate': 1.327884932081842, 'epoch': 0.02}                                                                                                                                         
{'loss': 10.9827, 'grad_norm': 180.7987518310547, 'learning_rate': 1.3125833022228446, 'epoch': 0.02}                                                                                                                                         
{'loss': 9.7317, 'grad_norm': 223.02679443359375, 'learning_rate': 1.2967264705660586, 'epoch': 0.02}                                                                                                                                         
{'loss': 11.4197, 'grad_norm': 147.71360778808594, 'learning_rate': 1.2803300858899105, 'epoch': 0.02}                                                                                                                                        
{'loss': 10.8513, 'grad_norm': 155.76853942871094, 'learning_rate': 1.2634103294465167, 'epoch': 0.02}                                                                                                                                        
{'loss': 9.7021, 'grad_norm': 114.98246002197266, 'learning_rate': 1.2459838989927388, 'epoch': 0.02}                                                                                                                                         
{'loss': 10.3884, 'grad_norm': 117.89942169189453, 'learning_rate': 1.2280679923115172, 'epoch': 0.02}                                                                                                                                        
{'loss': 8.0591, 'grad_norm': 132.5601348876953, 'learning_rate': 1.2096802902397323, 'epoch': 0.02}                                                                                                                                          
{'loss': 11.1552, 'grad_norm': 174.85458374023438, 'learning_rate': 1.1908389392193548, 'epoch': 0.02}                                                                                                                                        
{'loss': 11.5046, 'grad_norm': 273.67364501953125, 'learning_rate': 1.171562533389098, 'epoch': 0.03}                                                                                                                                         
{'loss': 10.6245, 'grad_norm': 146.13446044921875, 'learning_rate': 1.1518700962342474, 'epoch': 0.03}                                                                                                                                        
{'loss': 8.8656, 'grad_norm': 122.85801696777344, 'learning_rate': 1.1317810618127786, 'epoch': 0.03}                                                                                                                                         
{'loss': 10.1289, 'grad_norm': 124.23662567138672, 'learning_rate': 1.1113152555762864, 'epoch': 0.03}                                                                                                                                        
{'loss': 8.1793, 'grad_norm': 128.589599609375, 'learning_rate': 1.09049287480466, 'epoch': 0.03}                                                                                                                                             
{'loss': 8.302, 'grad_norm': 163.29751586914062, 'learning_rate': 1.0693344686738044, 'epoch': 0.03}                                                                                                                                          
{'loss': 10.8923, 'grad_norm': 306.87774658203125, 'learning_rate': 1.0478609179760854, 'epoch': 0.03}                                                                                                                                        
{'loss': 11.0958, 'grad_norm': 139.49423217773438, 'learning_rate': 1.0260934145135086, 'epoch': 0.03}                                                                                                                                        
{'loss': 10.3918, 'grad_norm': 193.5450897216797, 'learning_rate': 1.0040534401839687, 'epoch': 0.03}                                                                                                                                         
{'loss': 9.3784, 'grad_norm': 202.28379821777344, 'learning_rate': 0.9817627457812106, 'epoch': 0.03}                                                                                                                                         
{'loss': 9.7892, 'grad_norm': 130.73440551757812, 'learning_rate': 0.959243329529422, 'epoch': 0.03}                                                                                                                                          
{'loss': 9.0866, 'grad_norm': 251.4721221923828, 'learning_rate': 0.9365174153736413, 'epoch': 0.03}                                                                                                                                          
{'loss': 8.0445, 'grad_norm': 115.4397964477539, 'learning_rate': 0.913607431047407, 'epoch': 0.03}                                                                                                                                           
{'loss': 9.881, 'grad_norm': 157.1419219970703, 'learning_rate': 0.8905359859392936, 'epoch': 0.04}                                                                                                                                           
{'loss': 9.1954, 'grad_norm': 137.57525634765625, 'learning_rate': 0.8673258487801732, 'epoch': 0.04}                                                                                                                                         
{'loss': 8.0511, 'grad_norm': 160.0341796875, 'learning_rate': 0.8439999251732282, 'epoch': 0.04}                                                                                                                                             
{'loss': 9.062, 'grad_norm': 173.21034240722656, 'learning_rate': 0.8205812349888859, 'epoch': 0.04}                                                                                                                                          
{'loss': 8.0168, 'grad_norm': 195.0382537841797, 'learning_rate': 0.7970928896469851, 'epoch': 0.04}                                                                                                                                          
{'loss': 9.4908, 'grad_norm': 159.2719268798828, 'learning_rate': 0.7735580693085962, 'epoch': 0.04}                                                                                                                                          
{'loss': 10.3831, 'grad_norm': 117.14413452148438, 'learning_rate': 0.75, 'epoch': 0.04}                                                                                                                                                      
{'loss': 7.8719, 'grad_norm': 108.56270599365234, 'learning_rate': 0.7264419306914038, 'epoch': 0.04}                                                                                                                                         
{'loss': 8.8638, 'grad_norm': 171.00563049316406, 'learning_rate': 0.702907110353015, 'epoch': 0.04}                                                                                                                                          
{'loss': 9.3332, 'grad_norm': 180.0504913330078, 'learning_rate': 0.6794187650111143, 'epoch': 0.04}                                                                                                                                          
{'loss': 9.3431, 'grad_norm': 97.2490234375, 'learning_rate': 0.6560000748267717, 'epoch': 0.04}                                                                                                                                              
{'loss': 8.0484, 'grad_norm': 81.8037109375, 'learning_rate': 0.6326741512198267, 'epoch': 0.04}                                                                                                                                              
{'loss': 8.7684, 'grad_norm': 178.28945922851562, 'learning_rate': 0.6094640140607064, 'epoch': 0.05}                                                                                                                                         
{'loss': 9.1124, 'grad_norm': 142.69839477539062, 'learning_rate': 0.5863925689525933, 'epoch': 0.05}                                                                                                                                         
{'loss': 9.3765, 'grad_norm': 268.7080993652344, 'learning_rate': 0.563482584626359, 'epoch': 0.05}                                                                                                                                           
{'loss': 7.7957, 'grad_norm': 123.7179183959961, 'learning_rate': 0.5407566704705782, 'epoch': 0.05}                                                                                                                                          
{'loss': 8.7678, 'grad_norm': 189.27146911621094, 'learning_rate': 0.5182372542187895, 'epoch': 0.05}                                                                                                                                         
{'loss': 8.6127, 'grad_norm': 165.65679931640625, 'learning_rate': 0.4959465598160315, 'epoch': 0.05}                                                                                                                                         
{'loss': 9.6517, 'grad_norm': 118.67855834960938, 'learning_rate': 0.47390658548649156, 'epoch': 0.05}                                                                                                                                        
{'loss': 8.9602, 'grad_norm': 173.3568115234375, 'learning_rate': 0.4521390820239146, 'epoch': 0.05}                                                                                                                                          
{'loss': 9.4954, 'grad_norm': 109.3962631225586, 'learning_rate': 0.4306655313261955, 'epoch': 0.05}                                                                                                                                          
{'loss': 9.4025, 'grad_norm': 187.13485717773438, 'learning_rate': 0.40950712519533994, 'epoch': 0.05}                                                                                                                                        
{'loss': 9.1409, 'grad_norm': 121.71739196777344, 'learning_rate': 0.3886847444237134, 'epoch': 0.05}                                                                                                                                         
{'loss': 8.4716, 'grad_norm': 137.11294555664062, 'learning_rate': 0.3682189381872215, 'epoch': 0.05}                                                                                                                                         
{'loss': 9.4927, 'grad_norm': 155.65414428710938, 'learning_rate': 0.34812990376575237, 'epoch': 0.06}                                                                                                                                        
{'loss': 8.2565, 'grad_norm': 95.901123046875, 'learning_rate': 0.32843746661090223, 'epoch': 0.06}                                                                                                                                           
{'loss': 8.3361, 'grad_norm': 149.410400390625, 'learning_rate': 0.3091610607806452, 'epoch': 0.06}                                                                                                                                           
{'loss': 9.5941, 'grad_norm': 139.734619140625, 'learning_rate': 0.2903197097602678, 'epoch': 0.06}                                                                                                                                           
{'loss': 9.7985, 'grad_norm': 199.619384765625, 'learning_rate': 0.2719320076884827, 'epoch': 0.06}                                                                                                                                           
{'loss': 10.7871, 'grad_norm': 183.59854125976562, 'learning_rate': 0.25401610100726124, 'epoch': 0.06}                                                                                                                                       
{'loss': 8.6859, 'grad_norm': 112.32357788085938, 'learning_rate': 0.23658967055348346, 'epoch': 0.06}                                                                                                                                        
{'loss': 10.5418, 'grad_norm': 255.44549560546875, 'learning_rate': 0.2196699141100894, 'epoch': 0.06}                                                                                                                                        
{'loss': 8.7464, 'grad_norm': 218.6815185546875, 'learning_rate': 0.2032735294339415, 'epoch': 0.06}                                                                                                                                          
{'loss': 6.9212, 'grad_norm': 61.294891357421875, 'learning_rate': 0.1874166977771553, 'epoch': 0.06}                                                                                                                                         
{'loss': 9.0627, 'grad_norm': 114.74219512939453, 'learning_rate': 0.17211506791815814, 'epoch': 0.06}                                                                                                                                        
{'loss': 7.8808, 'grad_norm': 128.3472442626953, 'learning_rate': 0.1573837407182322, 'epoch': 0.06}                                                                                                                                          
{'loss': 9.6884, 'grad_norm': 175.79080200195312, 'learning_rate': 0.1432372542187895, 'epoch': 0.07}                                                                                                                                         
{'loss': 10.6194, 'grad_norm': 246.43626403808594, 'learning_rate': 0.12968956929407854, 'epoch': 0.07}                                                                                                                                       
{'loss': 9.3275, 'grad_norm': 123.32420349121094, 'learning_rate': 0.11675405587348886, 'epoch': 0.07}                                                                                                                                        
{'loss': 7.8527, 'grad_norm': 94.1834716796875, 'learning_rate': 0.10444347974704235, 'epoch': 0.07}                                                                                                                                          
{'loss': 6.564, 'grad_norm': 110.13874816894531, 'learning_rate': 0.09276998996710248, 'epoch': 0.07}                                                                                                                                         
{'loss': 8.8694, 'grad_norm': 111.09248352050781, 'learning_rate': 0.08174510685872416, 'epoch': 0.07}                                                                                                                                        
{'loss': 8.0952, 'grad_norm': 82.26622772216797, 'learning_rate': 0.07137971065048548, 'epoch': 0.07}                                                                                                                                         
{'loss': 10.9126, 'grad_norm': 222.2317352294922, 'learning_rate': 0.061684030737014145, 'epoch': 0.07}                                                                                                                                       
{'loss': 8.5527, 'grad_norm': 115.05807495117188, 'learning_rate': 0.05266763558381149, 'epoch': 0.07}                                                                                                                                        
{'loss': 7.8129, 'grad_norm': 127.63672637939453, 'learning_rate': 0.04433942328433091, 'epoch': 0.07}                                                                                                                                        
{'loss': 9.8106, 'grad_norm': 145.65557861328125, 'learning_rate': 0.03670761277863485, 'epoch': 0.07}                                                                                                                                        
{'loss': 8.2097, 'grad_norm': 129.25054931640625, 'learning_rate': 0.0297797357422927, 'epoch': 0.07}                                                                                                                                         
{'loss': 9.0377, 'grad_norm': 78.1057357788086, 'learning_rate': 0.023562629153526693, 'epoch': 0.07}                                                                                                                                         
{'loss': 9.0659, 'grad_norm': 86.17181396484375, 'learning_rate': 0.018062428545939424, 'epoch': 0.08}                                                                                                                                        
{'loss': 9.1692, 'grad_norm': 92.11599731445312, 'learning_rate': 0.013284561953483542, 'epoch': 0.08}                                                                                                                                        
{'loss': 8.8364, 'grad_norm': 75.83094024658203, 'learning_rate': 0.009233744553646756, 'epoch': 0.08}                                                                                                                                        
{'loss': 8.0378, 'grad_norm': 116.85031127929688, 'learning_rate': 0.005913974014141676, 'epoch': 0.08}                                                                                                                                       
{'loss': 9.115, 'grad_norm': 206.8473358154297, 'learning_rate': 0.0033285265476900028, 'epoch': 0.08}                                                                                                                                        
{'loss': 9.5083, 'grad_norm': 177.9818572998047, 'learning_rate': 0.0014799536787963308, 'epoch': 0.08}                                                                                                                                       
{'loss': 8.3315, 'grad_norm': 118.37013244628906, 'learning_rate': 0.0003700797257013, 'epoch': 0.08}                                                                                                                                         
{'loss': 8.6832, 'grad_norm': 166.12042236328125, 'learning_rate': 0.0, 'epoch': 0.08}                                                                                                                                                        
{'train_runtime': 2732.8754, 'train_samples_per_second': 0.037, 'train_steps_per_second': 0.037, 'train_loss': 9.700043871402741, 'epoch': 0.08}                                                                                              
100%|████████████████████████████████████| 100/100 [45:32<00:00, 27.33s/it]
```

![image](https://github.com/user-attachments/assets/a92708ce-86ff-40cc-9f58-bd4c1105ef59)

As we can see, the loss starts off with 3.0266, which is the loss when R1/R2 are Hadamard rotation matrices. But as the training steps forward, the loss boosts up at the second step and fails to converge within 100 steps. So will increasing the 'max_steps' help?

**Comment by limingwhat** [comment]
<img width="1356" alt="image" src="https://github.com/user-attachments/assets/fd31676a-c63d-418d-98bf-f5bf9c5f0fc1">

I did another two rounds of training, each 300 epochs. The blue curve is the first round. The orange curve is the second round, which use the optimized rotation matrices from round one as initial values. Both rounds start off with a very low loss about 3.0, then the losses boost up and stay around 6.0+, which I think is not a good loss value.

**Comment by ywlq** [comment]
@limingwhat @liuzechun 同样的问题，请问是如何解决的

**Comment by limingwhat** [comment]
> [@limingwhat](https://github.com/limingwhat) [@liuzechun](https://github.com/liuzechun) 同样的问题，请问是如何解决的

估计你使用的不是llama模型，需要注意一下和llama模型的区别。

**Comment by ywlq** [comment]
@limingwhat 对，我训练的是多模态的minicpm-v，整体是可以训练然后推理的，但是训练的过程的loss波动特别大，整体的loss也特别高，在1附近降不下来，想请问你训练的时候的loss大概是怎么下降的，能下降到什么程度，如果有Loss图就更好了
